### ISSUE TITLE
Bogus unnecessary_this with pattern matching and shadowing

### ISSUE THREAD

**Issue by nex3** [dart-lang/linter#4381]
For the code below:

```dart
class Foo {
  Object? value;

  bool equals(Object other) => switch (other) {
    Foo(: var value) => this.value == value,
    _ => false
  };
}
```

the 3.0.0 analyzer will print

```
info - test.dart:5:25 - Unnecessary 'this.' qualifier. Try removing 'this.'. - unnecessary_this
```

This is incorrect. The `this.` is necessary to disambiguate the containing class's `value` field from the `value` variable bound by the `switch` expression.

**Comment by pq** [comment]
Good catch. Taking a look now. Thanks!

**Comment by nex3** [comment]
@pq Is there an ETA on this ending up in a released Dart version? It looks like Dart 3.7.0 is still surfacing the bogus lint.

**Comment by pq** [comment]
My understanding is that a Dart 3.1 release is slated for the next few weeks. (@itsjustkevin could confirm?)
